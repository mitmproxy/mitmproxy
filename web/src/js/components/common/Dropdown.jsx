import React, {Component} from 'react'
import PropTypes from 'prop-types'
import classnames from 'classnames'

export const Divider = () => <hr className="divider"/>

export default class Dropdown extends Component {

    static propTypes = {
        dropup: PropTypes.bool,
        className: PropTypes.string,
        btnClass: PropTypes.string.isRequired
    }

    static defaultProps = {
        dropup: false
    }

    constructor(props, context) {
        super(props, context)
        this.state = {open: false}
        this.close = this.close.bind(this)
        this.open = this.open.bind(this)
    }

    close() {
        this.setState({open: false})
        document.removeEventListener('click', this.close)
    }

    open(e) {
        e.preventDefault()
        if (this.state.open) {
            return
        }
        e.stopPropagation();
        this.setState({open: !this.state.open})
        document.addEventListener('click', this.close)
    }

    render() {
        const {dropup, className, btnClass, text, children} = this.props
        return (
            <div className={classnames((dropup ? 'dropup' : 'dropdown'), className, {open: this.state.open})}>
                <a href='#' className={btnClass}
                   onClick={this.open}>
                    {text}
                </a>
                <ul className="dropdown-menu" role="menu">
                    {children.map((item, i) => <li key={i}> {item} </li>)}
                </ul>
            </div>
        )
    }
}
