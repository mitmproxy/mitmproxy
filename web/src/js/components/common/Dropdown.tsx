import React, { useEffect, useState } from "react";
import { useFloating, UseFloatingOptions } from "@floating-ui/react-dom";
import classnames from "classnames";

export const Divider = () => <li role="separator" className="divider" />;

type MenuItemProps = {
    onClick: () => void;
    children: React.ReactNode;
};

export function MenuItem({ onClick, children, ...attrs }: MenuItemProps) {
    const click = (e) => {
        e.preventDefault();
        onClick();
    };

    return (
        <li>
            <a href="#" onClick={click} {...attrs}>
                {children}
            </a>
        </li>
    );
}

type SubMenuProps = {
    title: string;
    children: React.ReactNode;
    className?: string;
};

export function SubMenu({ title, children, className }: SubMenuProps) {
    const [open, setOpen] = useState(false);
    const { refs, floatingStyles } = useFloating({
        placement: "right-start",
    });
    let submenu: React.ReactNode | null = null;
    if (open) {
        submenu = (
            <ul
                className={classnames("dropdown-menu show", className)}
                ref={refs.setFloating}
                style={floatingStyles}
            >
                {children}
            </ul>
        );
    }

    return (
        <li
            ref={refs.setReference}
            onMouseEnter={() => setOpen(true)}
            onMouseLeave={() => setOpen(false)}
        >
            <a>
                <i
                    className="fa fa-caret-right pull-right"
                    aria-hidden="true"
                />{" "}
                {title}
            </a>
            {submenu}
        </li>
    );
}

type DropdownProps = {
    text: React.ReactNode;
    children: React.ReactNode;
    options?: UseFloatingOptions;
    className?: string;
    onOpen?: (b: boolean) => void;
};
/*
 * When modifying this component, check that File -> Open and flow content upload work.
 */
export default React.memo(function Dropdown({
    text,
    children,
    options,
    className,
    onOpen,
    ...attrs
}: DropdownProps) {
    const [open, _setOpen] = useState(false);

    const { refs, floatingStyles } = useFloating(options);

    const setOpen = (b: boolean) => {
        _setOpen(b);
        if (onOpen) onOpen(b);
    };

    useEffect(() => {
        if (!refs.floating.current) return;
        document.addEventListener(
            "click",
            (e) => {
                if (!refs.floating.current?.contains(e.target as Node)) {
                    e.preventDefault();
                    e.stopPropagation();
                    setOpen(false);
                } else {
                    // We only want to call setOpen on the way out. This ensures the dropdown stays in DOM when clicked,
                    // which we need for file upload from Dropdowns.
                    document.addEventListener("click", () => setOpen(false), {
                        once: true,
                    });
                }
            },
            { once: true, capture: true },
        );
    }, [refs.floating.current]);

    let contents;
    if (open) {
        contents = (
            <ul
                className="dropdown-menu show"
                ref={refs.setFloating}
                style={floatingStyles}
            >
                {children}
            </ul>
        );
    } else {
        contents = null;
    }

    return (
        <>
            <a
                href="#"
                ref={refs.setReference}
                className={classnames(className, { open: open })}
                onClick={(e) => {
                    e.preventDefault();
                    setOpen(true);
                }}
                {...attrs}
            >
                {text}
            </a>
            {contents}
        </>
    );
});
