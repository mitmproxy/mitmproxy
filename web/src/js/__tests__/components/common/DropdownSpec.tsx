import * as React from "react";
import Dropdown, {
    Divider,
    MenuItem,
    SubMenu,
} from "../../../components/common/Dropdown";
import { fireEvent, render, screen, waitFor } from "../../test-utils";

test("Dropdown", async () => {
    const onOpen = jest.fn();
    const { asFragment } = render(
        <Dropdown text="open me" onOpen={onOpen}>
            <MenuItem onClick={() => 0}>click me</MenuItem>
            <Divider />
            <MenuItem onClick={() => 0}>click me</MenuItem>
        </Dropdown>,
    );
    expect(asFragment()).toMatchSnapshot();

    fireEvent.click(screen.getByText("open me"));
    await waitFor(() => expect(onOpen).toBeCalledWith(true));
    expect(asFragment()).toMatchSnapshot();

    onOpen.mockClear();
    fireEvent.click(document.body);
    await waitFor(() => expect(onOpen).toBeCalledWith(false));
});

test("SubMenu", async () => {
    const { asFragment } = render(
        <SubMenu title="submenu">
            <MenuItem onClick={() => 0}>click me</MenuItem>
        </SubMenu>,
    );
    expect(asFragment()).toMatchSnapshot();

    fireEvent.mouseEnter(screen.getByText("submenu"));
    await waitFor(() => screen.getByText("click me"));
    expect(asFragment()).toMatchSnapshot();

    fireEvent.mouseLeave(screen.getByText("submenu"));
    expect(screen.queryByText("click me")).toBeNull();
    expect(asFragment()).toMatchSnapshot();
});

test("MenuItem", async () => {
    const click = jest.fn();
    const { asFragment } = render(<MenuItem onClick={click}>wtf</MenuItem>);
    expect(asFragment()).toMatchSnapshot();
    fireEvent.click(screen.getByText("wtf"));
    await waitFor(() => expect(click).toBeCalled());
});
