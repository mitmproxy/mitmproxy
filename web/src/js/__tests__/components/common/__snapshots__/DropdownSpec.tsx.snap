// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown 1`] = `
<DocumentFragment>
  <a
    class=""
    href="#"
  >
    open me
  </a>
</DocumentFragment>
`;

exports[`Dropdown 2`] = `
<DocumentFragment>
  <a
    class="open"
    href="#"
  >
    open me
  </a>
  <ul
    class="dropdown-menu show"
    style="position: absolute; left: 0px; top: 0px; transform: translate(0px, 0px);"
  >
    <li>
      <a
        href="#"
      >
        click me
      </a>
    </li>
    <li
      class="divider"
      role="separator"
    />
    <li>
      <a
        href="#"
      >
        click me
      </a>
    </li>
  </ul>
</DocumentFragment>
`;

exports[`MenuItem 1`] = `
<DocumentFragment>
  <li>
    <a
      href="#"
    >
      wtf
    </a>
  </li>
</DocumentFragment>
`;

exports[`SubMenu 1`] = `
<DocumentFragment>
  <li>
    <a>
      <i
        aria-hidden="true"
        class="fa fa-caret-right pull-right"
      />
       submenu
    </a>
  </li>
</DocumentFragment>
`;

exports[`SubMenu 2`] = `
<DocumentFragment>
  <li>
    <a>
      <i
        aria-hidden="true"
        class="fa fa-caret-right pull-right"
      />
       submenu
    </a>
    <ul
      class="dropdown-menu show"
      style="position: absolute; left: 0px; top: 0px; transform: translate(0px, 0px);"
    >
      <li>
        <a
          href="#"
        >
          click me
        </a>
      </li>
    </ul>
  </li>
</DocumentFragment>
`;

exports[`SubMenu 3`] = `
<DocumentFragment>
  <li>
    <a>
      <i
        aria-hidden="true"
        class="fa fa-caret-right pull-right"
      />
       submenu
    </a>
  </li>
</DocumentFragment>
`;
